pkg/core/bundle_list.go

```diff
@@ -12,52 +13,139 @@ import (
 )
 
 const (
-	maxBundlesToList = 1000000
+	maxMetaFilesToProcess = 1000000
+	typicalBundlesNum     = 1000
 )
 
+func minInt(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// ListBundles returns the list of bundle descriptors from a repo.
+//
+// NOTE: resulting bundles are returned in no particular order.
+//
 // TODO: return a paginated list of id<->bd (separate function in repo_list.go)
-func ListBundles(repo string, store storage.Store) ([]model.BundleDescriptor, error) {
+func ListBundles(repo string, store storage.Store, opts ...BundleListOption) ([]model.BundleDescriptor, error) {
+	settings := CoreSettings{concurrentBundleList: defaultBundleListConcurrency}
+	for _, bApply := range opts {
+		bApply(&settings)
+	}
+
 	// Get a list
 	e := RepoExists(repo, store)
 	if e != nil {
 		return nil, e
 	}
-	ks, _, err := store.KeysPrefix(context.Background(), "", model.GetArchivePathPrefixToBundles(repo), "", maxBundlesToList)
+
+	// this call is not made async yet
+	ks, _, err := store.KeysPrefix(context.Background(), "", model.GetArchivePathPrefixToBundles(repo), "", maxMetaFilesToProcess)
 	if err != nil {
 		return nil, err
 	}
-	bds := make([]model.BundleDescriptor, 0)
-	for _, k := range ks {
+
+	var (
+		workers, wg sync.WaitGroup
+		werr        error
+	)
+
+	bundleChan := make(chan model.BundleDescriptor)
+	keyChan := make(chan string)
+	errorChan := make(chan error)
+
+	for i := 0; i < minInt(settings.concurrentBundleList, len(ks)); i++ {
+		workers.Add(1)
+		go getBundleAsync(repo, store, keyChan, bundleChan, errorChan, &workers)
+	}
+
+	bds := make([]model.BundleDescriptor, 0, typicalBundlesNum) // preallocate some typical number of bundles, e.g. 1000
+
+	wg.Add(1)
+	go func(bundleChan <-chan model.BundleDescriptor, wg *sync.WaitGroup) {
+		defer wg.Done()
+		for bd := range bundleChan { // watch for results and coalesce
```

Comment:
regarding the use of `select` mentioned on Slack, yes, there could be a

```
for {
	select {
		case bd, bundleChanMore := <- bundleChan:
			bds = append(bds, bd)
		case err := <- errChan:
			werr = err
	}
	if !bundleChanMore { 
		for {
			err, errChanMore := <- errChan
			if err != nil { werr = err }
			if !errChanMore { break }
		}
		break
	}
}
```

or some similar at the end of `ListBundles`.  with this, there's no need for the `WaitGroup` declared as `wg`.  closing the `bundleChan` and `errorChan` (in that order) is detected via the [receive operator](https://golang.org/ref/spec#Receive_operator).